design/resource-dependency.md

```diff
@@ -0,0 +1,348 @@
+
+### Generic core engine consideration
+
+The design of resource dependency is a pretty generic engine that works for any k8s resources.
+Even though we use OAM native fields like `Parameters` or `Outputs` fields, those are just representation layer.
+The underlying dependency engine along with the DataInput and DataOutput model is the core implementation,
+which should be generic enough to convert or adopt future concepts or fields.
+
+### Spec upgrade path consideration
+
+Initially the feature will be experimental. It might change iteratively.
+Once that happens, we strive to keep it upgradeable between two releases.
+Here we discuss an idea to upgrade if that happens.
+
+The core dependency engine in OAM runtime is a generic implementation.
+We would map APIs to underlying DataInput/DataOutput objects.
+
+It's inevitable that an intermediate phase where both old and new APIs co-exist.
+But they don't both exist in the same ApplicationConfiguration.
+Only one could exist and we don't need to handle any conflicts.
+In that case, We could add an webhook to translate the old to the new API.
+
+## Solution examples
+
+In this section we are providing more solution examples for dependency use cases.
+
+1. Web app and database.
+
+    ```yaml
+    kind: Component
+    metadata:
+      name: my-rds
+    spec:
+      workload:
+        kind: RDSInstance
+        metadata:
+          name: my-rds
+    ---
+
+    kind: Component
+    metadata:
+      name: my-web
+    spec:
+      workload:
+        kind: MyWorkload
+        metadata:
+          name: my-web
+      parameters:
+      - name: database-secret
+        fieldPaths: ["spec.connSecret"]
+    ---
+
+    kind: ApplicationConfiguration
+    spec:
+      components:
+      - componentName: my-web
+        parameterValues:
+        - name: database-secret
+          valueFrom:
+            dataOutputName: mysql-conn-secret
+      - componentName: my-rds
+        dataOutputs:
+          - name: mysql-conn-secret
+            fieldPath: "status.connSecret"
+    ```
+
+2. Service binding trait.
+
+    ApplicationConfiguration :
+
+    ```yaml
+    kind: ApplicationConfiguration
+    spec:
+      components:
+      - componentName: my-web
+        traits:
+        - trait:
+            apiVersion: core.oam.dev/v1alpha1
+            kind: ServiceBinding
+            metadata:
+              name: my-secret-binding
+            spec:
+              from:
+                secret:
+                  name: my-secret
+              to:
+                env: true
+          dataOutputs:
+            - name: my-secret-binding
+              fieldPath: status.ready
+    ```
+
+    Component:
+
+    ```yaml
+    kind: Component
+    metadata:
+      name: my-web
+    spec:
+      parameters:
+      - name: my-secret-binding
+        fieldPaths: [] # This parameter is for ordering purpose only
```

Comment:
this seems a bit hacky as there is no data passing

Reply:
synced with @hongchaodeng offline.  I get that this is a work-around for pure dependency cases. My take is that since OAM adopts the declarative model, one doesn't really have to enforce this type of dependency as it will play out at the end. The controllers should just give up reconcile for a few times before they eventually succeed